File: Personal/CommandPattern/pkg/document_invoker.py
from pkg.document import Document
from pkg.document_write_command import DocumentWriteCommand
from pkg.document_erase_command import DocumentEraseCommand


class DocumentInvoker:
    """Document invoker class..."""

    def __init__(self, name):
        """Constructor.

        Parameters
        ----------
        name : str
        """
        self._commands = []
        self._current_doc = Document(name)

    def write(self, text):
        """Instantiate 'DocumentWriteCommand' and store the write command in '_commands'.

        Parameters
        ----------
        text : str
            Text to be added to the the document.
        """
        cmd = DocumentWriteCommand(self._current_doc, text)
        self._commands.append(cmd)

    def erase(self, text):
        """Instantiate 'DocumentEraseCommand' and store the erase command in '_commands'.

        Parameters
        ----------
        text : str
            Text to be added to the the document.
        """
        cmd = DocumentEraseCommand(self._current_doc, text)
        self._commands.append(cmd)

    def read(self):
        """Call the read() method inside of the 'Document' class and return the provided
        text.

        Returns
        -------
        str
            Combined text from '_text_array' inside of the 'Document' class.
        """
        return self._current_doc.read()

    def undo(self, index):
        """Undo the actions of a command index 'index'.

        Parameters
        ----------
        index : int
            The index of a command inside of '_commands'.
        """
        tmp = self._commands[index]
        print("\n---- Undo level {}".format(index))
        tmp.undo()

    def redo(self, index):
        """Undo the actions of the 'undo' method.

        Parameters
        ----------
        index : int
            The index of a command inside of '_commands'.
        """
        tmp = self._commands[index]
        print("\n---- Redo level {}".format(index))
        tmp.redo()

    def print_command_buffer(self):
        """Return the content of '_commands'.

        Returns
        -------
        txt : str
            A string comprised of all the text in '_commands'.
        """
        txt = "\n------Commands issued------"

        for line in self._commands:
            txt += "\n" + str(line)

        return txt

File: CSCD212/CommandPattern/src/DocumentInvoker.java
import java.util.ArrayList;

public class DocumentInvoker {
    // Keeps a history of every command used.
    private ArrayList<Command> commands = new ArrayList<>();
    private Document           currentDoc;

    /**
     * Constructor: instantiates a Document object.
     * @param name Name of the new document.
     */
    public DocumentInvoker(String name) { currentDoc = new Document(name); }

    /**
     * Instantiate a 'DocumentWriteCommand' object to add 'text' to the document.
     * @param text Text to be added to the document.
     */
    public void Write(String text) {
        DocumentWriteCommand cmd = new DocumentWriteCommand(currentDoc, text);
        // Add the used command to array.
        commands.add(cmd);
    }

    /**
     * Instantiate a 'DocumentWriteCommand' object to erase 'text' from the document.
     * @param text Text to be erased from the document.
     */
    public void Erase(String text) {
        DocumentEraseCommand cmd = new DocumentEraseCommand(currentDoc, text);
        commands.add(cmd);
    }

    /**
     * Instantiate a 'DocumentBoldCommand' object to bold the text at 'index' in the
     * document.
     * @param line Index of text to be bolded.
     */
    public void Bold(int line) {
        DocumentBoldCommand cmd = new DocumentBoldCommand(currentDoc, line);
        commands.add(cmd);
    }

    /**
     * Instantiate a 'DocumentRemoveBoldCommand' object to un-bold the text at 'index'
     * in the document.
     * @param line Index of text to be un-bolded.
     */
    public void RemoveBold(int line) {
        DocumentRemoveBoldCommand cmd = new DocumentRemoveBoldCommand(currentDoc, line);
        commands.add(cmd);
    }

    /**
     * Undo the undo...
     * @param level Index...
     */
    public void Redo(int level) {
        Command tmp;
        System.out.println("---- Redo level " + level);
        tmp = commands.get(level);
        tmp.redo();
    }

    /**
     * Undo a previous action.
     * @param level Index...
     */
    public void Undo(int level) {
        Command tmp;
        System.out.println("---- Undo level " + level);
        tmp = commands.get(level);
        tmp.undo();
    }

    /** ... */
    public String Read() { return currentDoc.Read(); }

    /** ... */
    public void printCommmandBuffer() {
        StringBuffer txt = new StringBuffer();

        txt.append("----Commands issued-----");
        for (int i = 0; i < commands.size(); i++) {
            txt.append(i).append(":").append(commands.get(i)).append("\n");
        }
        System.out.println(txt);
    }
}
